stellargraph/layer/deep_graph_infomax.py

```diff
@@ -127,14 +128,22 @@ def build(self):
             GlobalAveragePooling1D()(self._node_feats)
         )
 
-        discriminator = BilinearTransform()
+        discriminator = DGIDiscriminator(name=self._unique_id)
```

Comment:
Why set the name on this layer rather than on the layer that is actually used by `embedding_model`?

Reply:
unfortunately `_node_feats` is a tensor not a layer, so I don't know of a straightforward way of renaming the relevant layer.

Reply:
Ah, right. So this `DGIDiscriminator` layer basically ends up as a flag marking a model as coming from this `DeepGraphInfoMax` instance, and then, having done that validation, we pull out the actual embedding layer separately?

I'm a little concerned by the mutation with `self._node_feats`: if someone creates two models using a single instance, I think they'll get different models for training, but a very confused embedding model (I think the out tensor will be the one from the second model, always):

```python
dgi = DeepGraphInfoMax(...)
inp1, out1 = dgi.build()
model1 = Model(inpt1, out1)

inp2, out2 = dgi.build()
model2 = Model(inpt2, out2)

emb_inp1, emb_out1 = dgi.embedding_model(model1)
emb_inp2, emb_out2 = dgi.embedding_model(model2)
```

Given this, I suspect `emb_out1 is emb_out2` (i.e. they're exact the same object) and is the embedding layer for `model2`, but `emb_inp1` are the inputs for `model1`. That is, the embedding layer for `model1` has been lost.

Maybe, to address all of this, we could put the marker on the embedding layer by creating a pass-through identity layer and not doing any mutations of the `DeepGraphInfoMax` object, something like:

```python
x_inp, x_out = self.base_model.build(multiplicity=1)

# identity layer so that we can name it to retrieve in `embedding_model`
node_feat = Lambda(lambda x: x, shape=x_out.shape[1:], name=...)(x_out)
```

(Inspired by https://stackoverflow.com/a/47297842/1256624)

Reply:
hmm yeah that's a good point. I'll wrap `node_feats` in a `Lambda` layer with a randomly generated name